picard/tagger.py

```diff
@@ -851,7 +852,7 @@ def _scan_paths_recursive(paths, recursive, ignore_hidden):
             except OSError as err:
                 log.warning(err)
 
-    def add_paths(self, paths, target=None):
+    async def add_paths(self, paths, target=None):
```

Comment:
I don't get why this is introduced at this point. Can you explain why it is needed?

Additionally `add_paths()` is very likely to be used by external plugins, making it a coroutine will break them.

It should be noted that asyncio wasn't used at all in the code until its introduction here. I don't say we shouldn't use asyncio but I fear it breaks things.

@phw what do you think?

Reply:
I really did not want to introduce asyncio to the codebase but I could not come up with a better idea to actually **await** for the files to be loaded. If you have any better idea how to run commands after all the files are already loaded, I’ll gladly remove the asyncio.

Reply:
If i just tried adding some thread (pool) and block the function until the amount of all files in picard matches the number of arguments, I ended up with the whole app getting frozen

Reply:
> I really did not want to introduce asyncio to the codebase but I could not come up with a better idea to actually **await** for the files to be loaded

That's fine, I have nothing against using asyncio if it helps, I'm just worried those changes could break something.

Reply:
Actually, seems like the issue persists even with asyncio (but there's some improvement). I suppose that we might need to write some blocking file-loading flow for the sake of `picard -e LOAD` command in the next PR

Reply:
So @phw, the question for you is whether we merge at the current state or reset to https://github.com/metabrainz/picard/pull/2137/commits/4d0867faaef54d21bbde3a8d784a44029c6946df and re-apply the `picard -e SAVE_MATCHED` patch, while leaving the `asyncio` stuff for a next ticket

Reply:
After further research: I tend to reset to the mentioned commit. We need to introduce second type of all executed commands: blocking ones. This is too much for this PR imo. 

Even though we await load, there is still save, lookup, scan etc. Let's leave it for now.

Reply:
We do not want to run GUI events in one thread, but the command-line commands should be executed in a blocking way to ensure proper execution order. 

So my diagnosis is:
- each new command requires research if it uses any threaded code. If yes, it has to be cloned without this threaded code (or the futures have to be joined at some place).

And my suggestion is:
- reset to https://github.com/metabrainz/picard/commit/4d0867faaef54d21bbde3a8d784a44029c6946df, re-apply `picard -e SAVE_MATCHED` patch and merge
- in the second PR let's take care of the execution order: i.e. join threads & await functions in their clone versions for our CLI purposes.
- in the third one let's add:
    - `LOAD`
    - `LOG_DUMP`
    - `LOG_CLEAR`
    - `FROM_FILE`: (load commands from a given text file)
    
Sorry for the mess but it's really complex to me, I'm still learning the codebase.

What's your take on this @phw @zas?

Reply:
I agree with your suggestion.